v4.0.1-alpha.2

Spacesai1or opened the pull request:
### Added

#### web3-eth-contract

-   Decoding error data, using Error ABI if available, if error was returned from a smart contract function call (#5662).

#### web3-types

-   These types were moved from `web3-eth-accounts` to `web3-types` package: Cipher, CipherOptions, ScryptParams, PBKDF2SHA256Params, KeyStore (#5581 )

#### web3-utils

-   Export a new function `uuidV4` that generates a random v4 Uuid (#5373).
-   Enable passing a starting number, to increment based on it, for the Json Rpc Request `id` (#5652).
-   Export a new function `isPromise` that checks if an object is a promise (#5652).

#### web3-eth-contract

-   `SpecialOutput` type was added as a generic type into the call function to support reassigning output types (#5631)
-   Overloaded functions types (`ContractOverloadedMethodInputs`, `ContractOverloadedMethodOutputs`) was added (#5631)

### Fixed

#### web3-eth-contract

-   Emit past contract events based on `fromBlock` when passed to `contract.events.someEventName` (#5201)
-   Use different types for `ContractOptions` -> `jsonInterface` setter and getter (#5474)
-   An issue within the `Contract` constructor where `provider` wasn't being set when provided within the `optionsOrContextOrReturnFormat` argument (#5669)

#### web3-types

-   Make the `request` method of `EIP1193Provider` class, compatible with EIP 1193 (#5591)

#### web3-utils

-   Use Uuid for the response id, to fix the issue "Responses get mixed up due to conflicting payload IDs" (#5373).

#### web3-validator

-   Fix `isHex`returning `false` for `-123`, fix `isHexStrict` returning `true` for `-0x`, and fix `isHex` returning `true` for empty strings `` (#5373).

#### web3-eth-abi

-   Fix ContractMethodOutputParameters type to support output object types by index and string key. Also, it returns void if ABI doesn't have outputs and returns exactly one type if the output array has only one element. (#5631)

### Removed

#### web3-eth-accounts

-   These types were moved from `web3-eth-accounts` to `web3-types` package: Cipher, CipherOptions, ScryptParams, PBKDF2SHA256Params, KeyStore (#5581 )

Spacesai1or committed:
Version bump and build for v4.0.1-alpha.2

jdevcs commented on tools/web3-packagetemplate/package.json:
```diff
@@ -1,6 +1,6 @@
 {
 	"name": "web3-packagetemplate",
-	"version": "0.1.1-alpha.0",
+	"version": "0.1.1-alpha.1",
```
I think there isn't any change in this package ?, so we dnt need version bump.  Its easier to review if we include changelog changes in same PR.

Spacesai1or committed:
Release/4.0.1 alpha.2 changelog update 3 (#5689)

* Update CHANGELOG version headers

* Update CHANGELOGs

* Revert package version bump for web3-packagetemplate

avkos commented:
@spacesailor24 thanks for PR. just need to merge 4.x into this branch and resolve conflicts.

Spacesai1or commented:
> @spacesailor24 thanks for PR. just need to merge 4.x into this branch and resolve conflicts.

@avkos So usually we would have an RC version we could bump every time we make changes to the release since the last commit in the release is tagged with the version. We've already messed up by merging #5689 into this branch, since now the branch has commits that come after the tagged commit. This isn't a huge issue since this only really affects the continuity of the [release listed in Github](https://github.com/web3/web3.js/releases/tag/untagged-e2800057933dc0d92b7d) (I'm talking about the `Assets` listed in the release). I believe the right steps would've been to open a PR that merges `4.x` into the release branch to solve the conflicts and cut another RC, but we don't have that option now since there are no RC versions for `4.x` yet

So, if we are to merge `4.x` into the release branch and solve the conflicts, we'd need to re-build all the packages, commit, tag the commit, and re-create the release on Github

jdevcs commented:
> > @spacesailor24 thanks for PR. just need to merge 4.x into this branch and resolve conflicts.
> 
> @avkos So usually we would have an RC version we could bump every time we make changes to the release since the last commit in the release is tagged with the version. We've already messed up by merging #5689 into this branch, since now the branch has commits that come after the tagged commit. This isn't a huge issue since this only really affects the continuity of the [release listed in Github](https://github.com/web3/web3.js/releases/tag/untagged-e2800057933dc0d92b7d) (I'm talking about the `Assets` listed in the release). I believe the right steps would've been to open a PR that merges `4.x` into the release branch to solve the conflicts and cut another RC, but we don't have that option now since there are no RC versions for `4.x` yet
> 
> So, if we are to merge `4.x` into the release branch and solve the conflicts, we'd need to re-build all the packages, commit, tag the commit, and re-create the release on Github

In case of merge conflicts, and before actual release, could you go ahead with following:
Merge main branch into release PR, build, delete old tag, recreate new tag, and re populate github draft release with newly created tag. There isn't any technical limitation or issue in that, if it happens before release on npm / github. 

In future we should also follow freezing main branch for 4.x if there is release PR created, we were following this in 1.x.

Spacesai1or committed:
Release/4.0.1 alpha.2 merge conflicts (#5692)

* Update CHANGELOG version headers

* Contract call with tuple is missing param names (#5613)

* call special output type

* fix

* fix: enable outputs to have param names (#5624)

* hot fix

* add type if only one param

* overloaded inputs types

* fix resolver tests

* add type tests

* simplify types

* revert registry unit test

* test firefox

* revert firefox test

* update changelogs

Co-authored-by: Marin Petrunić <mpetrunic@users.noreply.github.com>

* Fix `isHex` and `isHexStrict` for some edge cases and enrich their tests (#5655)

* Change `isHex` to return true for negative numbers (for example `-123`)
* Change both `isHex` and `isHexStrict` to return `false` for `-0x`
* Change `isHex` to return `false` for empty strings ''.

* Remove erroneous set provider code for Contract constructor (#5669)

* Remove erroneous set provider code for Contract constructor. Add Contract constructor provider set test

* Update CHANGELOGs

* Debugging failing integration tests

* Apply suggestions from code review

* Use getSystemTestProvider instead of hardcoded string

* Refactor ternaries in Contract constructor to if statements

* Correct CHANGELOG entries

* Remove unneeded checks in if statements

* Test with injected external providers (#5652)

* fix: sending tx with injected provider (#5651)

Co-authored-by: Marin Petrunic <marin.petrunic@gmail.com>

* adding a test for using `ganache` provider

* enable the jsonrpc `id` to optionally be incremented starting from a number
 (Inspired by: https://github.com/web3/web3.js/issues/5373#issuecomment-1227073060 and needed as a work around for https://github.com/blockchainsllc/in3/issues/46)

* test with `in3` as a provider & skip `in3` test if it takes too long

* increase integration test timeout at web3 package

* add a test for using `hardhat` provider

* improve how legacy providers, such as hardhat, is used

* implement `isPromise` that works in the browsers

* refactor external providers tests

* update CHANGELOG.md

* Use Error ABI to parse errors when sending a transaction (#5662)

* use Error ABI when sending a transaction

* update CHANGELOG.md for #5587

* Remove merge conflict marker

Co-authored-by: Oleksii Kosynskyi <oleksii.kosynskyi@gmail.com>
Co-authored-by: Marin Petrunić <mpetrunic@users.noreply.github.com>
Co-authored-by: Muhammad Altabba <24407834+Muhammad-Altabba@users.noreply.github.com>
Co-authored-by: jdevcs <junaid@chainsafe.io>

jdevcs committed:
Merge branch '4.x' into release/4.0.1-alpha.2

jdevcs commented on packages/web3-eth-abi/CHANGELOG.md:
```diff
@@ -46,6 +46,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+
+-   Updated Web3.js dependencies (#5664)
+
 ## [Unreleased]
```
This should be at end of file

jdevcs commented on packages/web3-validator/CHANGELOG.md:
```diff
@@ -41,6 +41,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+
+-   Updated Web3.js dependencies (#5664)
+
 ## [Unreleased]
```
This should be at end of file

Spacesai1or committed:
Release patch/4.0.1 alpha.2/solve merge conflicts (#5697)

* Update CHANGELOG version headers

* Remove merge conflict markers

* Remove double CHANGELOG entry
